src/assembly/manifest.ts

```diff
@@ -0,0 +1,166 @@
+     * a file within the directory with the DEFAULT_FILENAME
+     */
+    public static fromPath(filePath: string): AssemblyManifestReader {
+        let st;
+        try {
+            st = fs.statSync(filePath);
+        } catch (e: any) {
+            throw new Error(`Cannot read manifest at '${filePath}': ${e.message}`);
+        }
+        if (st.isDirectory()) {
+            return AssemblyManifestReader.fromFile(path.join(filePath, AssemblyManifestReader.DEFAULT_FILENAME));
+        }
+        return AssemblyManifestReader.fromFile(filePath);
+    }
+
+    /**
+     * The directory where the manifest was found
+     */
+    public readonly directory: string;
+
+    private readonly assembly: CloudAssembly;
+    private readonly stacks = new Map<string, CloudFormationStackArtifact>();
+    private readonly _stackManifests = new Map<string, StackManifest>();
+    private readonly tree: ConstructTree;
+
+    constructor(directory: string, private readonly manifest: AssemblyManifest) {
+        this.directory = directory;
+        this.assembly = new CloudAssembly(directory, {
+            // we don't need version checking / version checking would mean we would have to
+            // publish a new version of the library everytime the cdk version increases
+            skipVersionCheck: true,
+        });
+        this.tree = fs.readJsonSync(path.resolve(this.directory, 'tree.json')).tree;
+        if (!this.tree.children) {
+            throw new Error('Invalid tree.json found');
+        }
+        this.renderStackManifest();
+    }
+
+    private renderStackManifest() {
+        for (const [artifactId, artifact] of Object.entries(this.manifest.artifacts ?? {})) {
+            if (artifact.type === ArtifactType.AWS_CLOUDFORMATION_STACK) {
+                const stackArtifact = this.assembly.getStackArtifact(artifactId);
+                this.stacks.set(artifactId, stackArtifact);
+                const metadata: StackMetadata = {};
+                const props = artifact.properties as AwsCloudFormationStackProperties;
+                const template: CloudFormationTemplate = fs.readJSONSync(
+                    path.resolve(this.directory, props.templateFile),
+                );
+                for (const [metadataId, metadataEntry] of Object.entries(artifact.metadata ?? {})) {
+                    metadataEntry.forEach((meta) => {
+                        if (meta.type === ArtifactMetadataEntryType.LOGICAL_ID) {
+                            // For some reason the metadata entry prefixes the path with a `/`
+                            const path = metadataId.startsWith('/') ? metadataId.substring(1) : metadataId;
+                            metadata[path] = meta.data as LogicalIdMetadataEntry;
+                        }
+                    });
+                }
+                const assets = this.getAssetsForStack(artifactId);
+                const stackTree = this.tree.children![artifactId];
```

Comment:
This non-null assertion should be fine because we're doing this check in the calling method, but it might still be good to add one here too be safe for the future.

Reply:
added